### PR TITLE
[Query] Allow Client Eval in Projection

### LINKED ISSUE #15715: Query: Materialize an array
Current binding is member based and does not support arrays. Hence projection gets overwritten.

### LINKED ISSUE #15716: Query: Enable client eval in projection

### LINKED ISSUE #15719: Query: NewExpression.Members can be null
Perhaps binding similar to array can help but I doubt without the member infos we can compose any further.

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #15715
Resolves #15719
Resolves #15716

### CODE REVIEW

File: src/EFCore.Relational/Query/Pipeline/SqlExpressions/SelectExpression.cs [hunk 1]
```diff
@@ -103,6 +104,41 @@ public SelectExpression SetProjectionAsResult(SqlExpression sqlExpression)
...
+        }
+
+        public Expression GetProjectionExpression(ProjectionMember projectionMember)
+        {
+            return _projectionMapping[projectionMember];
+        }
+
+        public ProjectionBindingExpression AddToProjection(SqlExpression sqlExpression, Type type)
+        {
+            _projection.Add(new ProjectionExpression(sqlExpression, ""));
```
Comment (maumar) [inline src/EFCore.Relational/Query/Pipeline/SqlExpressions/SelectExpression.cs]: nit: alias: "" for readability, same goes in other places

File: src/EFCore.Relational/Query/Pipeline/RelationalShapedQueryCompilingExpressionVisitor.cs [hunk 2]
```diff
@@ -269,9 +269,12 @@ protected override Expression VisitBinary(BinaryExpression binaryExpression)
                     && parameterExpression.Type == typeof(MaterializationContext))
                 {
                     var newExpression = (NewExpression)binaryExpression.Right;
+                    var projectionBindingExpression = (ProjectionBindingExpression)newExpression.Arguments[0];
 
                     _materializationContextBindings[parameterExpression]
-                        = (int)((ConstantExpression)_selectExpression.GetProjectionExpression(((ProjectionBindingExpression)newExpression.Arguments[0]).ProjectionMember)).Value;
+                        = projectionBindingExpression.ProjectionMember != null
```
Comment (maumar) [inline src/EFCore.Relational/Query/Pipeline/RelationalShapedQueryCompilingExpressionVisitor.cs]: repeated 3 times - DRY

File: src/EFCore.Relational/Query/Pipeline/RelationalProjectionBindingExpressionVisitor.cs [hunk 3]
```diff
@@ -131,17 +188,33 @@ protected override Expression VisitNew(NewExpression newExpression)
         protected override Expression VisitMemberInit(MemberInitExpression memberInitExpression)
         {
             var newExpression = (NewExpression)Visit(memberInitExpression.NewExpression);
```
Comment (maumar) [inline src/EFCore.Relational/Query/Pipeline/RelationalProjectionBindingExpressionVisitor.cs]: newNewExpression ;P

File: src/EFCore.Relational/Query/Pipeline/RelationalProjectionBindingExpressionVisitor.cs [hunk 4]
```diff
@@ -64,65 +74,112 @@ public override Expression Visit(Expression expression)
...
 
-                return new ProjectionBindingExpression(_selectExpression, _projectionMembers.Peek(), expression.Type);
+                    return new ProjectionBindingExpression(_selectExpression, _projectionMembers.Peek(), expression.Type);
+                }
             }
 
             return base.Visit(expression);
         }
 
+        private static readonly MethodInfo _getParameterValueMethodInfo
```
Comment (maumar) [inline src/EFCore.Relational/Query/Pipeline/RelationalProjectionBindingExpressionVisitor.cs]: is this the right place for this method?

Comment (smitpatel) [reply src/EFCore.Relational/Query/Pipeline/RelationalProjectionBindingExpressionVisitor.cs]: For now let's keep it here. It is used internally and not in core yet. We can DRY up code later if we see it is being used across providers.

**Pull request merged**